crates/services/txpool_v2/src/collision_manager/mod.rs

```diff
@@ -45,15 +49,20 @@ impl<Idx> Collisions<Idx> {
     }
 }
 
-pub trait CollisionManager<S: Storage> {
+pub trait CollisionManagerStorage {
+    type StorageIndex: Copy + Debug;
+
+    fn get(&self, index: &Self::StorageIndex) -> Result<&StorageData, Error>;
+}
+
+pub trait CollisionManager<S: CollisionManagerStorage> {
```

Comment:
We talked about maybe putting the generic `S` on the trait methods instead of on the trait proper. I think this will help clean up `impl` blocks around the code without giving up much.

Reply:
The problem is that if I place the genrics on the methods I can't use the assiociated type definition because it's not specified in the trait.

Trait def: 
![Screenshot 2024-09-12 141258](https://github.com/user-attachments/assets/c6180fe1-2d17-4374-b92d-36f1197f6a0c)

Impl:
![Screenshot 2024-09-12 141314](https://github.com/user-attachments/assets/959525a0-4225-4c47-b754-8335acfc54b4)

Compile error:
![Screenshot 2024-09-12 141334](https://github.com/user-attachments/assets/ce152e20-26d4-4934-9d08-a2d13abefb4d)

Reply:
I've been thinking a bit about this and I don't see any obvious better way to do it.

TL; DR - I think we can remove the generics completely from the trait, but it would require an extra wrapper type over the current collision manager which feels a bit overkill.

**Why I don't think we should put generics on the methods**

Putting the generics on the methods themselves makes it a bit weird because that would assume the same collision manager could be called multiple times with different storage implementations. I have a hard time reasoning about the full implications of that but it sounds counter-intuitive and as @AurelienFT already pointed out it prevents reusing the associated storage index type. I'm happy to discuss this suggestion further if you still think it makes sense @MitchTurner. 

**How I ideally would like the relation between this and the storage trait to look**

I think the main problem here is the coupling between the collision manager logic and the storage trait. I think the cleanest thing would be to omit the storage details from the collision manager trait, as the storage connection should be the concern of the implementing struct.

```
pub trait CollisionManager {
    type StorageIndex: Copy + Debug;
    
    fn collect_colliding_transactions(
        &self,
        transaction: &PoolTransaction,
    ) -> Result<Collisions<Self::StorageIndex>, Error>;

    fn on_stored_transaction(
        &mut self,
        transaction: &PoolTransaction,
        transaction_id: Self::StorageIndex,
    ) -> Result<(), Error>;

    fn on_removed_transaction(
        &mut self,
        transaction: &PoolTransaction,
    ) -> Result<(), Error>;
}
```

However this would be tedious in practice because `BasicCollisionManager` still needs to have a reference to the storage to implement these methods. Naively putting a reference in `BasicCollisionManager` wouldn't work, because then we wouldn't be able to mutate the storage as long as there is a `BasicCollisionManager` in scope. The overkill solution would be to wrap it in an ephemeral struct which implements the `CollisionManager` trait. Something like

```
struct BasicCollisionManagerWithStorageRef<'a, Storage> {
    inner: BasicCollisionManager,
    ref: &'a Storage,
}
```

And implement the trait on the wrapper struct. But that feels like over-doing it and I can respect the current solution as good enough.

Reply:
How about:
```rs
pub trait CollisionManager {
    type Storage: CollisionManagerStorage;

    fn collect_colliding_transactions(
        &self,
        transaction: &PoolTransaction,
        storage: &Self::Storage,
    ) -> Result<Collisions<<Self::Storage as CollisionManagerStorage>::StorageIndex>, Error>;

    fn on_stored_transaction(
        &mut self,
        transaction: &PoolTransaction,
        transaction_id: <Self::Storage as CollisionManagerStorage>::StorageIndex,
    ) -> Result<(), Error>;

    fn on_removed_transaction(
        &mut self,
        transaction: &PoolTransaction,
    ) -> Result<(), Error>;
}
```
and
```rs
pub struct BasicCollisionManager<StorageIndex, S> {
    messages_spenders: HashMap<Nonce, StorageIndex>,
    coins_spenders: HashMap<UtxoId, StorageIndex>,
    contracts_creators: HashMap<ContractId, StorageIndex>,
    blobs_users: HashMap<BlobId, StorageIndex>,
    _storage: std::marker::PhantomData<S>,
}
```
and
```rs
impl<StorageIndex, S> CollisionManager for BasicCollisionManager<StorageIndex, S>
where
    StorageIndex: Copy + Debug,
    S: CollisionManagerStorage<StorageIndex = StorageIndex>,
{
    type Storage = S;
    
    ...
}
```

I got that to compile :)

The main thing I'm avoiding here is the consumer of `trait CollisionManager` needing to know anything about `trait CollisionManagerStorage`. And also saving the trait interface of `CollisionManager` from being cluttered with extra information.

This seems to achieve both of those things 👍

Reply:
I like it. It's not perfect, but it is an improvement :+1: . My main concern about using an associated type for the storage is that it constrains the implementer to only ever work with a single type of storage (i.e., what if I want to implement CollisionManager for type X and have the implementation support multiple types of storage). However, we get around this limitation here by having the `BasicCollisionManager` type be generic over the storage type instead, and I suppose this would be the case for any collisionmanager we implement and in the worst case we can always add phantom data like above to fix this.

Reply:
>My main concern about using an associated type for the storage is that it constrains the implementer to only ever work with a single type of storage (i.e., what if I want to implement CollisionManager for type X and have the implementation support multiple types of storage).

Well, I never committed to any specific `CollisionManagerStorage`, instead I just implement `CollisionManager` generically for all `S: CollisionManagerStorage`. That will be populated at compile time. We only ever need to test the `CollisionManager` interface of `BasicCollisionManager` anyway, so we can pass it a fake `CollisionManagerStorage` in tests with no trouble :)

Reply:
Okay. I think this might be even better:
```
pub trait CollisionManager {
    type Storage;
    type StorageIndex;

    fn collect_colliding_transactions(
        &self,
        transaction: &PoolTransaction,
        storage: &Self::Storage,
    ) -> Result<Collisions<<Self::StorageIndex>, Error>;

    fn on_stored_transaction(
        &mut self,
        transaction: &PoolTransaction,
        transaction_id: <Self::StorageIndex,
    ) -> Result<(), Error>;

    fn on_removed_transaction(
        &mut self,
        transaction: &PoolTransaction,
    ) -> Result<(), Error>;
}
```
and
```
impl<S> CollisionManager for BasicCollisionManager<S::StorageIndex, S>
where
    S: CollisionManagerStorage,
{
    type Storage = S;
    type StorageIndex = S::StorageIndex;
    
    ...
}
```
Then we are truly decoupling `CollisionManager` from `CollisionManagerStorage` :)

Reply:
That's true, my point is that that only works if `BasicCollisionManager` is generic. The constraint we're accepting here is that if we were to implement `CollisionManager` for a non-generic type we'd have to commit to a specific `CollisionManagerStorage`. Since we're not doing that I think this is okay :)

Reply:
Yeah that's even better :rocket:

Reply:
I highly prefer this one : https://github.com/FuelLabs/fuel-core/pull/2162#discussion_r1758467612 . My interrogation is, why do we want to decouple that much the `CollisionManagerStorage` and `CollisionManager` ?

This forces us to duplicate tthe associated type `StorageIndex` in both traits and enforce that they are the same only at implementation level.
For me they will always be the same, that's the point of `CollisionManagerStorage` (it's like a port for the storage in the context of the `CollisionManager`) it's to serve the data for the `CollisionManager`. I don't see any reason why an other would be the data provider in the future.

Reply:
And rename `CollisionManagerStorage` to `BasicCollisionManagerStorage` and you could do:

```
pub struct BasicCollisionManager<S: BasicCollisionManagerStorage> {
    messages_spenders: HashMap<Nonce, S::StorageIndex>,
    coins_spenders: HashMap<UtxoId, S::StorageIndex>,
    contracts_creators: HashMap<ContractId, S::StorageIndex>,
    blobs_users: HashMap<BlobId, S::StorageIndex>,
}
```

Reply:
>This forces us to duplicate tthe associated type StorageIndex in both traits and enforce that they are the same only at implementation level.

Fixed that :) for the struct at least.

> For me they will always be the same, that's the point of CollisionManagerStorage (it's like a port for the storage in the context of the CollisionManager) it's to serve the data for the CollisionManager. I don't see any reason why an other would be the data provider in the future.

True, but we only need to communicate that in one place:
```rs
impl<PS, View, S, CM, SA> Pool<PS, S, CM, SA>
where
    PS: AtomicView<LatestView = View>,
    View: TxPoolPersistentStorage,
    S: Storage,
    S: CollisionManagerStorage<StorageIndex = <S as Storage>::StorageIndex>,
    S: SelectionAlgorithmStorage<StorageIndex = <S as Storage>::StorageIndex>,
    CM: CollisionManager<Storage = S, StorageIndex = <S as Storage>::StorageIndex>,
    SA: SelectionAlgorithm<S>,
{ 
    ...
}
```
In every other situation it doesn't matter that they are the same. The entire reason I suggested we create the `CollisionManagerStorage` was it was dependent on the implementation of `CollisionManager`. This communicates that finally without tying it to `CollisionManager` at all.

Reply:
My point is that, if someone come to the codebase and check the traits and see that `CollisionManager` and `CollisionManagerStorage` are not tied at all this can lead to misundersttanding in my mind as they are meant to work together.

> True, but we only need to communicate that in one place:

With the current implementation it already communicate this :
```rust
impl<PS, View, S, CM, SA> Pool<PS, S, CM, SA>
where
    PS: AtomicView<LatestView = View>,
    View: TxPoolPersistentStorage,
    S: Storage,
    S: CollisionManagerStorage<StorageIndex = <S as Storage>::StorageIndex>,
    S: SelectionAlgorithmStorage<StorageIndex = <S as Storage>::StorageIndex>,
    CM: CollisionManager<S>,
    SA: SelectionAlgorithm<S>,
{
```

Reply:
After internal discussions, we agreed that this solution is the best one. It's implemented on this PR. @MitchTurner if everything is ok for you you can resolve :)